### DISCUSSION TITLE
How to use smartcard to logon rdp ?

### DISCUSSION THREAD

**Discussion by YeyeX64** [FreeRDP/FreeRDP#9436] [Q&A] [answered]
hello, i try to use smartcard to logon on rdp windows logon.

The smartcard is detected by windows, the card too but when i put the PIN, windows write "driver is not initialized".

Has anyone ever encountered the problem?

**Comment by akallabeth** [comment]
@YeyeX64 no, but your description lacks quite a lot of details ;)
adding the *exact*  steps might help in answering your question.

**Reply by YeyeX64** [reply, marked as answer]
Hello, sorry.

On my client, debian10. FreeRDP version 2.11.1.
Smartcard and the card reader seem to be ok :

`tcos@openthin10:~$ opensc-tool --list-readers
# Detected readers (pcsc)
Nr.  Card  Features  Name
0    No              CCID SC Reader (076B:A022) 00 00
tcos@openthin10:~$ opensc-tool --list-readers
# Detected readers (pcsc)
Nr.  Card  Features  Name
0    Yes             CCID SC Reader (076B:A022) 00 00
`

On widows session rdp, the card is detected because when i insert it in card reader, windows ask pin code.
But when i do the PIN, widows say that driver necessary for the card is not initialized.

Do you need windows logs ?

Best regards

**Reply by hardening** [reply]
Better use 3.0 for smartcard logon, it's not functional in 2.x

**Reply by akallabeth** [reply]
@YeyeX64 that sounds like you have tested `smartcard redirection` ? Smartcard logon is quite a different matter (and as @hardening already told, it requires `freerdp 3`)
also, please what does `widows say that driver necessary for the card is not initialized.` mean? how did you test that?

so, to debug this in a meaningful manner:
1. try `certutil -SCInfo` to test your normal smartcard redirection in a running `RDP` session
2. check if you can actually use the certificates on your smartcard for `RDP` login with `mstsc` from a windows machine
3. only now try to use the `/smartcard-logon` argument of `xfreerdp` to try to log on

**Comment by hardening** [comment]
FYI, I wrote a post on things to check to do smartcard logon with FreeRDP 3.0: https://www.hardening-consulting.com/en/posts/20231004-smartcard-logon.html

**Comment by YeyeX64** [comment]
> @hardening 
> 1. try `certutil -SCInfo` to test your normal smartcard redirection in a running `RDP` session

via RDP normal logon (without card ans PIN) : 
`certutil -SCInfo
Le gestionnaire de ressource des cartes à puce est en cours d’exécution.
État de la carte/lecteur actuel :
Lecteurs : 1
  0: SCM Microsystems Inc. SCR 355 [CCID Interface] 00 00
--- Lecteur : SCM Microsystems Inc. SCR 355 [CCID Interface] 00 00
--- Statut : SCARD_STATE_PRESENT | SCARD_STATE_INUSE
--- Statut : La carte est partagée par un autre processus.
---   Carte : Carte de Professionnel de Sante CPS3
---    ATR :
        3b ac 00 40 2a 00 12 25  00 64 80 00 03 10 00 90   ;..@*..%.d......
        00                                                 .


=======================================================
Analyse de la carte dans le lecteur : SCM Microsystems Inc. SCR 355 [CCID Interface] 00 00

--------------===========================--------------
================ Certificat 0 ================
--- Lecteur : SCM Microsystems Inc. SCR 355 [CCID Interface] 00 00
---   Carte : Carte de Professionnel de Sante CPS3
Fournisseur = ASIP Sante Cryptographic Provider
Conteneur de clé = 010UFR1$0"UPyrnes-Atlantiques(64)10UXXXXXXXXXXXXXXX10U16407808211#0!UPersonnelsantousocial1>0U*XXXX [Conteneur par défaut]

Numéro de série : 421f4a2b66345a3bb82f00676b12da6b
Émetteur: CN=AC IGC-SANTE STANDARD PERSONNES, OU=IGC-SANTE, OU=XXX, O=ASIP-SANTE, C=FR
 NotBefore : 23/01/2023 22:26
 NotAfter : 23/01/2026 22:26
Objet: G=XXX + SN=XXX + CN=XXXX, T=Personnel santé ou social, OU=XXX, O=XXX, S=Pyrénées-Atlantiques (64), C=FR
Il ne s’agit pas d’un certificat racine
Hach. cert. (sha1) : c64633c7ef3fd82f2652a6815b7e03d4aaecf63a

Exécution du test de correspondance de clé publique AT_SIGNATURE...
Test de correspondance de clé publique réussi
  Conteneur de clé = \\.\SCM Microsystems Inc. SCR 355 [CCID Interface] 00 00\010UFR1$0"UPyrnes-Atlantiques(64)10UXXXX10U16407808211#0!UPersonnelsantousocial1>0U*XXX0U3640780821/XXX
  Fournisseur = ASIP Sante Cryptographic Provider
  ProviderType = 1
  Indicateurs = 1
    0x1 (1)
  KeySpec = 2 -- AT_SIGNATURE
La clé privée vérifie

Vérification de la chaîne de certificat en cours...
La clé privée vérifie

Vérification de la chaîne de certificat en cours...
CertGetCertificateChain(dwErrorStatus) = 0x1000040
Chaîne de carte à puce non valide
dwFlags = CA_VERIFY_FLAGS_CONSOLE_TRACE (0x20000000)
dwFlags = CA_VERIFY_FLAGS_DUMP_CHAIN (0x40000000)
ChainFlags = CERT_CHAIN_REVOCATION_CHECK_CHAIN_EXCLUDE_ROOT (0x40000000)
HCCE_LOCAL_MACHINE
CERT_CHAIN_POLICY_BASE
-------- CERT_CHAIN_CONTEXT --------
ChainContext.dwInfoStatus = CERT_TRUST_HAS_PREFERRED_ISSUER (0x100)
ChainContext.dwErrorStatus = CERT_TRUST_REVOCATION_STATUS_UNKNOWN (0x40)
ChainContext.dwErrorStatus = CERT_TRUST_IS_OFFLINE_REVOCATION (0x1000000)

SimpleChain.dwInfoStatus = CERT_TRUST_HAS_PREFERRED_ISSUER (0x100)
SimpleChain.dwErrorStatus = CERT_TRUST_REVOCATION_STATUS_UNKNOWN (0x40)
SimpleChain.dwErrorStatus = CERT_TRUST_IS_OFFLINE_REVOCATION (0x1000000)

CertContext[0][0]: dwInfoStatus=102 dwErrorStatus=1000040
  Issuer: CN=AC IGC-SANTE STANDARD PERSONNES, OU=IGC-SANTE, OU=0002 187512751, O=ASIP-SANTE, C=FR
  NotBefore: 23/01/2023 22:24
  NotAfter: 23/01/2026 22:24
  Subject: G=OLIVIER + SN=xxx+ CN=3640xxx780821/xxx, T=Personnel santé ou social, OU=xxx, O=xxx, S=Pyrénées-Atlantiques (64), C=FR
  Serial: 58d05935f8619857b25fa9470b6310da
  SubjectAltName: Autre nom :Nom principal=xxx.xxxx@carte-cps.fr
  Cert: 6d7aba149f3f4590b2624dca7e66405814922828
  Element.dwInfoStatus = CERT_TRUST_HAS_KEY_MATCH_ISSUER (0x2)
  Element.dwInfoStatus = CERT_TRUST_HAS_PREFERRED_ISSUER (0x100)
  Element.dwErrorStatus = CERT_TRUST_REVOCATION_STATUS_UNKNOWN (0x40)
  Element.dwErrorStatus = CERT_TRUST_IS_OFFLINE_REVOCATION (0x1000000)
  Issuance[0] = 1.2.250.1.213.1.7.1.2.1.3.1
  Application[0] = 1.3.6.1.5.5.7.3.2 Authentification du client
  Application[1] = 1.3.6.1.4.1.311.20.2.2 Ouverture de session par carte à puce

CertContext[0][1]: dwInfoStatus=102 dwErrorStatus=1000040
  Issuer: CN=AC RACINE IGC-SANTE STANDARD, OU=IGC-SANTE, OU=0002 187512751, O=ASIP-SANTE, C=FR
  NotBefore: 25/06/2013 02:00
  NotAfter: 24/06/2033 02:00
  Subject: CN=AC IGC-SANTE STANDARD PERSONNES, OU=IGC-SANTE, OU=0002 187512751, O=ASIP-SANTE, C=FR
  Serial: 112016654c36f13671b030541f6088ea44c2
  Cert: e0d6e1d8481fed62c65c53ed9dfcda949c3d1d81
  Element.dwInfoStatus = CERT_TRUST_HAS_KEY_MATCH_ISSUER (0x2)
  Element.dwInfoStatus = CERT_TRUST_HAS_PREFERRED_ISSUER (0x100)
  Element.dwErrorStatus = CERT_TRUST_REVOCATION_STATUS_UNKNOWN (0x40)
  Element.dwErrorStatus = CERT_TRUST_IS_OFFLINE_REVOCATION (0x1000000)

CertContext[0][2]: dwInfoStatus=10a dwErrorStatus=0
  Issuer: CN=AC RACINE IGC-SANTE STANDARD, OU=IGC-SANTE, OU=0002 187512751, O=ASIP-SANTE, C=FR
  NotBefore: 25/06/2013 02:00
  NotAfter: 25/06/2033 02:00
  Subject: CN=AC RACINE IGC-SANTE STANDARD, OU=IGC-SANTE, OU=0002 187512751, O=ASIP-SANTE, C=FR
  Serial: 11207f69865cdf9a675f3013640fe1a47805
  Cert: b6ba1d6d5224bceda95a67f6f37b86896edd0006
  Element.dwInfoStatus = CERT_TRUST_HAS_KEY_MATCH_ISSUER (0x2)
  Element.dwInfoStatus = CERT_TRUST_IS_SELF_SIGNED (0x8)
  Element.dwInfoStatus = CERT_TRUST_HAS_PREFERRED_ISSUER (0x100)

Exclude leaf cert:
  Chain: b6b832e50d19806199d20bdc48b111e5013c91aa
Full chain:
  Chain: 228cde55c54fe8075fb5f8934ed465167acd0d56
  Issuer: CN=AC IGC-SANTE STANDARD PERSONNES, OU=IGC-SANTE, OU=0002 187512751, O=ASIP-SANTE, C=FR
  NotBefore: 23/01/2023 22:24
  NotAfter: 23/01/2026 22:24
  Subject: G=OLIVIER + SN=xxxx+ CN=3xxx, T=Personnel santé ou social, OU=1640780821, O=xxxx SAINTE MARIE, S=Pyrénées-Atlantiques (64), C=FR
  Serial: 58d05935f8619857b25fa9470b6310da
  SubjectAltName: Autre nom :Nom principal=xxxxx@carte-cps.fr
  Cert: xxxxxx
La fonction de révocation n’a pas pu vérifier la révocation car le serveur de révocation était déconnecté. 0x80092013 (-2146885613 CRYPT_E_REVOCATION_OFFLINE)
------------------------------------
Vérification de révocation ignorée -- le serveur est hors connexion
Certificat AT_KEYEXCHANGE affiché pour le lecteur : Cherry KC 1000 SC 0
SCardGetCardTypeProviderName: Le fichier spécifié est introuvable. 0x2 (WIN32: 2 ERROR_FILE_NOT_FOUND)
Impossible de récupérer le nom de fournisseur pour Carte de Professionnel de Sante CPS3
--------------===========================--------------

Terminé.
CertUtil: -SCInfo La commande s’est terminée correctement.
`


> 2. check if you can actually use the certificates on your smartcard for `RDP` login with `mstsc` from a windows machine

Craps ... I hadn't thought about trying this.
So i use a rdp with enablecredsspsupport:i:0 parameter and same "behavior", rdp close.

> 3. only now try to use the `/smartcard-logon` argument of `xfreerdp` to try to log on

i will try soon with freerdp 3
